clis/teliod/src/main.rs

```diff
@@ -129,19 +194,169 @@ impl CommandListener {
+    // tests with core API. This is to not look for tokens in a test environment
+    // right now as the values are dummy and program will not run as it expects
+    // real tokens.
+    if !client_config.auth_token.eq("") {
+        start_telio(&mut telio, client_config.private_key)?;
+        if let Err(e) = update_meshmap(&client_config, &telio) {
+            error!("Unable to set meshmap due to {e}");
+        }
+
+        while let Some(cmd) = rx_channel.blocking_recv() {
+            info!("Got command {:?}", cmd);
+            match cmd {
+                TelioTaskCmd::UpdateMeshmap => {
+                    if let Err(e) = update_meshmap(&client_config, &telio) {
+                        error!("Unable to set meshmap due to {e}");
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn update_meshmap(client_config: &ClientConfig, telio: &Device) -> Result<(), TeliodError> {
+    let meshmap: MeshMap = serde_json::from_str(&get_meshmap(client_config)?)?;
+    trace!("Meshmap {:#?}", meshmap);
+    telio.set_config(&Some(meshmap))?;
+    Ok(())
+}
+
+fn start_telio(telio: &mut Device, private_key: SecretKey) -> Result<(), DeviceError> {
+    if !telio.is_running() {
+        let device_config = DeviceConfig {
+            private_key,
+            name: Some("utun10".to_owned()),
+            adapter: AdapterType::BoringTun,
+            ..Default::default()
+        };
+
+        telio.start(&device_config)?;
+
+        info!(
+            "started telio with {:?}:{}...",
+            AdapterType::BoringTun,
+            private_key
+        );
+    }
+    Ok(())
+}
+
+impl ClientConfig {
+    async fn new(config: &mut TeliodDaemonConfig) -> Result<ClientConfig, TeliodError> {
+        info!("Fetching tokens");
+
+        // Copy or generate Secret/public key pair
+        let private_key = if let Some(sk) = config.private_key {
+            sk
+        } else {
+            debug!("Generating secret key!");
+            SecretKey::gen()
+        };
+        let public_key = private_key.public();
+
+        // Copy or generate hw_identifier
+        let hw_identifier = if let Some(hw_id) = &config.hw_identifier {
+            hw_id.to_string()
+        } else {
+            debug!("Generating hw identifier");
+            format!("{}.{}", public_key, "openWRT")
+        };
+
+        // Copy or generate machine_identifier
+        let machine_identifier = if let Some(machine_id) = &config.machine_identifier {
+            machine_id.to_string()
+        } else {
+            if let Ok(id) = load_identifier_from_api(&config.authentication_token, public_key).await
+            {
+                id
+            } else {
+                debug!("Machine not yet registered");
+                // If registering fails. Close the daemon
+                register_machine(&hw_identifier, public_key, &config.authentication_token).await?
```

Comment:
You are right. I think registering here can be skipped completely and after initializing tokens go to updating the machine. It would return error as identifier is invalid, and simply register the machine then.